### PR TITLE
OP-1035 - Patient Consensus Management

### PR DESCRIPTION
**Pull request by goto-eof**
See OP-1035.

Paired with https://github.com/informatici/openhospital-core/pull/905.

### CODE REVIEW

File: src/main/java/org/isf/patconsensus/mapper/PatientConsensusMapper.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package org.isf.patconsensus.mapper;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patconsensus/mapper/PatientConsensusMapper.java:1]: Missing license header

File: src/main/java/org/isf/patconsensus/mapper/PatientConsensusMapper.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
+/*
+
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patconsensus/mapper/PatientConsensusMapper.java:2]: how did this blank line get into the license header block?

**Comment by mwithi** [comment]
I've tried with postman collection, I'm not able to update the consensus, it says patientcode mismatch

image

**Comment by goto-eof** [comment]
> I've tried with postman collection, I'm not able to update the consensus, it says patientcode mismatch
> 
> image

Can't reproduce the bug

Screenshot from 2023-04-02 12-29-06

**Review by mwithi — CHANGES_REQUESTED** [review]
The PatientConsensus creation should be done in the POST patient endpoint

**Comment by goto-eof** [comment]
> The PatientConsensus creation should be done in the POST patient endpoint

done

File: src/main/java/org/isf/patient/mapper/PatientMapper.java [hunk 3]
```diff
@@ -70,6 +71,7 @@ public Patient map2Model(PatientDTO toObj) {
 			photo.setPhoto(toObj.getBlobPhoto());
 			patient.setPatientProfilePhoto(photo);
 		}
+		patient.setPatientConsensus(new PatientConsensus(toObj.isConsensusFlag(), toObj.isConsensusAdministrativeFlag(), toObj.isConsensusServiceFlag(), patient));
```
Comment (mwithi) [inline src/main/java/org/isf/patient/mapper/PatientMapper.java:74]: Until we don't have the "ui" ready for sending the patient consensus flags, we should force to true the first one in the PatientController

File: src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import io.swagger.annotations.Api;
+import io.swagger.annotations.Authorization;
+
+@RestController()
+@RequestMapping("/patient-consensus")
+@Api(value = "/patient-consensus", produces = MediaType.APPLICATION_JSON_VALUE, authorizations = { @Authorization(value = "apiKey") })
```
Comment (mwithi) [inline src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java:53]: Let's use standard naming convention for api

```suggestion
@RequestMapping("/patientConsensus")
@Api(value = "/patientConsensus", produces = MediaType.APPLICATION_JSON_VALUE, authorizations = { @Authorization(value = "apiKey") })
```

File: src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	@Autowired
+	protected PatientConsensusMapper mapper;
+
+	@GetMapping(value = "/{patientId}", produces = MediaType.APPLICATION_JSON_VALUE)
+	ResponseEntity<PatientConsensusDTO> getPatientConsensus(@PathVariable Integer patientId) throws OHServiceException {
+		LOGGER.info("Retrieving patient consensus: {}", patientId);
+		Optional<PatientConsensus> patientConsensus = manager.getPatientConsensusByUserId(patientId);
+		if (patientConsensus.isEmpty()) {
+			return ResponseEntity
+							.status(HttpStatus.NO_CONTENT).body(null);
```
[the remark this answers is not in the document]
Comment (goto-eof) [reply src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java]: fixed, thanks

File: src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+		PatientConsensusDTO patientDTO = mapper.map2DTO(patientConsensus.get());
+		return ResponseEntity.ok(patientDTO);
+	}
+
+	@PutMapping(value = "/{patientId}", produces = MediaType.APPLICATION_JSON_VALUE)
+	ResponseEntity<PatientConsensusDTO> updatePatientConsensus(@PathVariable Integer patientId, @RequestBody PatientConsensusDTO patientConsensus)
+					throws OHServiceException {
+		LOGGER.info("Update patient consensus by id: {}", patientId);
+		if (patientId != patientConsensus.getPatientId()) {
+			throw new OHAPIException(new OHExceptionMessage(null, "Patient code mismatch", OHSeverityLevel.ERROR));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java:78]:
```suggestion
			throw new OHAPIException(new OHExceptionMessage("Patient code mismatch."));
```
Use new constructor.

File: src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+		}
+		Optional<PatientConsensus> patConsensusOpt = this.manager.getPatientConsensusByUserId(patientId);
+		if (patConsensusOpt.isEmpty()) {
+			throw new OHAPIException(new OHExceptionMessage(null, "PatientConsensus not found!", OHSeverityLevel.ERROR));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java:82]:
```suggestion
			throw new OHAPIException(new OHExceptionMessage("PatientConsensus not found."));
```
New constructor and don't finish messages with `!`

File: src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+		}
+		PatientConsensus updatedPatienConsensustModel = mapper.map2Model(patientConsensus);
+		updatedPatienConsensustModel.setId(this.manager.getPatientConsensusByUserId(patientId).get().getId());
+		PatientConsensus patientConsensusUpdated = manager.updatePatientConsensus(updatedPatienConsensustModel);
+		if (patientConsensusUpdated == null) {
+			throw new OHAPIException(new OHExceptionMessage(null, "PatientConsensus is not updated!", OHSeverityLevel.ERROR));
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java:88]:
```suggestion
			throw new OHAPIException(new OHExceptionMessage("PatientConsensus is not updated."));
```
same comments as above

File: src/main/java/org/isf/patient/mapper/PatientMapper.java [hunk 9]
```diff
@@ -49,13 +50,17 @@ public PatientDTO map2DTO(Patient fromObj) {
 	@Override
 	public PatientDTO map2DTOWS(Patient fromObj, Boolean status) {
 		PatientDTO patientDTO = super.map2DTOWS(fromObj, status);
-		if (fromObj.getPatientProfilePhoto()!= null) {
+		if (fromObj.getPatientProfilePhoto() != null) {
 			try {
 				patientDTO.setBlobPhoto(fromObj.getPatientProfilePhoto().getPhoto());
 			} catch (Exception e) {
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patient/mapper/PatientMapper.java:57]: why is the catch block empty?  Doesn't seem correct.

File: src/main/java/org/isf/patient/rest/PatientController.java [hunk 10]
```diff
@@ -90,6 +90,10 @@ public PatientController(PatientBrowserManager patientManager, AdmissionBrowserM
 	ResponseEntity<PatientDTO> newPatient(@RequestBody PatientDTO newPatient) throws OHServiceException {
 		String name = StringUtils.hasLength(newPatient.getName()) ? newPatient.getFirstName() + " " + newPatient.getSecondName() : newPatient.getName();
 		LOGGER.info("Create patient {}", name);
+
+		// XXX this flag is mandatory
```
Comment (mwithi) [inline src/main/java/org/isf/patient/rest/PatientController.java:94]:
```suggestion
		//TODO: remove this line when UI will be ready to collect the patient consensus
```

File: src/main/java/org/isf/patient/mapper/PatientMapper.java [hunk 11]
```diff
@@ -49,13 +54,17 @@ public PatientDTO map2DTO(Patient fromObj) {
...
-
+				LOGGER.error("error while trying to retrieve the profile photo");
```
Comment (mwithi) [inline src/main/java/org/isf/patient/mapper/PatientMapper.java]:
```suggestion
				LOGGER.error("Error while trying to retrieve the profile photo.");
```

File: src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+		}
+		PatientConsensusDTO patientDTO = mapper.map2DTO(patientConsensus.get());
+		return ResponseEntity.ok(patientDTO);
+	}
+
+	@PutMapping(value = "/{patientId}", produces = MediaType.APPLICATION_JSON_VALUE)
+	ResponseEntity<PatientConsensusDTO> updatePatientConsensus(@PathVariable Integer patientId, @RequestBody PatientConsensusDTO patientConsensus)
+					throws OHServiceException {
+		LOGGER.info("Update patient consensus by id: {}", patientId);
+		if (patientId != patientConsensus.getPatientId()) {
```
Comment (mwithi) [inline src/main/java/org/isf/patconsensus/rest/PatientConsensusController.java:77]:
```suggestion
		if (!patientId.equals(patientConsensus.getPatientId())) {
```

File: src/main/java/org/isf/patient/mapper/PatientMapper.java [hunk 13]
```diff
@@ -24,38 +24,50 @@
...
 
 	@Override
 	public PatientDTO map2DTO(Patient fromObj) {
 		PatientDTO patientDTO = super.map2DTO(fromObj);
-		if (fromObj.getPatientProfilePhoto() != null) {
-			patientDTO.setBlobPhoto(fromObj.getPatientProfilePhoto().getPhoto());
-		}
+//		if (fromObj.getPatientProfilePhoto() != null) {
+//			patientDTO.setBlobPhoto(fromObj.getPatientProfilePhoto().getPhoto());
+//		}
```
Comment (mwithi) [inline src/main/java/org/isf/patient/mapper/PatientMapper.java:57]: we don't need commented code, we can still find it in the history

File: src/main/java/org/isf/patient/mapper/PatientMapper.java [hunk 14]
```diff
@@ -24,38 +24,50 @@
...
-
-			}
-		}
+//		if (fromObj.getPatientProfilePhoto() != null) {
+//			try {
+//				patientDTO.setBlobPhoto(fromObj.getPatientProfilePhoto().getPhoto());
+//			} catch (Exception e) {
+//				LOGGER.error("Error while trying to retrieve the profile photo.");
+//			}
+//		}
```
Comment (mwithi) [inline src/main/java/org/isf/patient/mapper/PatientMapper.java:70]: we don't need commented code, we can still find it in the history

File: src/main/java/org/isf/patient/mapper/PatientMapper.java [hunk 15]
```diff
@@ -64,12 +76,12 @@ public PatientDTO map2DTOWS(Patient fromObj, Boolean status) {
...
-			photo.setPatient(patient);
-			photo.setPhoto(toObj.getBlobPhoto());
-			patient.setPatientProfilePhoto(photo);
-		}
+//		if (toObj.getBlobPhoto() != null) {
+//			PatientProfilePhoto photo = new PatientProfilePhoto();
+//			photo.setPatient(patient);
+//			photo.setPhoto(toObj.getBlobPhoto());
+//			patient.setPatientProfilePhoto(photo);
+//		}
```
Comment (mwithi) [inline src/main/java/org/isf/patient/mapper/PatientMapper.java:84]: we don't need commented code, we can still find it in the history

File: src/main/java/org/isf/patient/mapper/PatientMapper.java [hunk 16]
```diff
@@ -24,53 +24,30 @@
...
-
-	}
-	@Override
-	public PatientDTO map2DTOWS(Patient fromObj, Boolean status) {
-		PatientDTO patientDTO = super.map2DTOWS(fromObj, status);
-		if (fromObj.getPatientProfilePhoto()!= null) {
-			try {
-				patientDTO.setBlobPhoto(fromObj.getPatientProfilePhoto().getPhoto());
-			} catch (Exception e) {
+	private static final Logger LOGGER = org.slf4j.LoggerFactory.getLogger(PatientMapper.class);
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patient/mapper/PatientMapper.java:41]: why is this added yet never used?

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 17]
```diff
@@ -30,6 +30,7 @@
 import static org.hamcrest.Matchers.containsString;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.when;
+import org.isf.shared.mapper.mappings.PatientMapping;
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/patient/rest/PatientControllerTest.java:33]: Why is there a non-static import in the middle of the `import static` lines?

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java [hunk 18]
```diff
@@ -142,16 +150,15 @@ public void when_post_patients_is_call_with_empty_body_then_BadRequest_HttpMessa
...
-						post(request)
-								.content(empty_body.getBytes())
-								.contentType(MediaType.APPLICATION_JSON)
-				)
-				.andDo(log())
-				.andExpect(status().is4xxClientError())
-				.andExpect(status().isBadRequest())
-				.andExpect(content().string(anyOf(nullValue(), equalTo(""))))
-				.andReturn();
+						.perform(
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/patient/rest/PatientControllerTest.java:153]: This indentation seems excessive....

Comment (goto-eof) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java:153]: I used the indentation of OH, but I change it manualy

Comment (goto-eof) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java:153]: There are too many occurencies, can we modify the OH indentation file settings?

Comment (goto-eof) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java:153]: Screenshot from 2023-04-21 21-58-34

File: src/main/java/org/isf/patient/rest/PatientController.java [hunk 19]
```diff
@@ -78,9 +84,17 @@ public PatientController(PatientBrowserManager patientManager, AdmissionBrowserM
 		this.admissionManager = admissionManager;
 		this.patientMapper = patientMapper;
 	}
+	
+	public PatientController(PatientBrowserManager patientManager, AdmissionBrowserManager admissionManager, PatientMapper patientMapper, PatientConsensusBrowserManager patientConsensusManager) {
+		this.patientManager = patientManager;
+		this.admissionManager = admissionManager;
+		this.patientMapper = patientMapper;
+		this.patientConsensusManager = patientConsensusManager;
+	}
```
Comment (mwithi) [inline src/main/java/org/isf/patient/rest/PatientController.java:93]: it seems that to have two constructors confuses the ApplicationContext:
```
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
[02/may./2023 18:51:44] [:] ERROR - Application run failed
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'patientController' defined in file [D:\IDE\workspace.2022-03\openhospital-api\target\classes\org\isf\patient\rest\PatientController.class]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.isf.patient.rest.PatientController]: No default constructor found; nested exception is java.lang.NoSuchMethodException: org.isf.patient.rest.PatientController.<init>()
```

Do we really need to add a second constructor rather than modifying the former?
By commenting the new constructor application runs well (test not, of course) and all endpoints tests are ok (using Postman). Also using current UI (develop) is ok! 👏

**Review by mwithi — CHANGES_REQUESTED** [review]
remove/fix PatientController constructur

**Review by mwithi — APPROVED** [review]

File: src/test/java/org/isf/patient/rest/PatientControllerTest.java (hunk 18 above)
Comment (mwithi) [reply src/test/java/org/isf/patient/rest/PatientControllerTest.java:153]: I tried to format too that part and I don't see changes, looks like GitHub spacing looks different but the code is correct

image

**Review by mwithi — APPROVED** [review]
Latest changes work for me! Thanks a lot!
